azure-iot-device/azure/iot/device/common/error_map_paho.py

```diff
@@ -0,0 +1,110 @@
+    mqtt.CONNACK_REFUSED_PROTOCOL_VERSION: errors_paho.PahoConnectionRefusedProtocolVersionError,
+    mqtt.CONNACK_REFUSED_IDENTIFIER_REJECTED: errors_paho.PahoConnectionRefusedIdentifierRejectedError,
+    mqtt.CONNACK_REFUSED_SERVER_UNAVAILABLE: errors_paho.PahoConnectionRefusedServerUnavailableError,
+    mqtt.CONNACK_REFUSED_BAD_USERNAME_PASSWORD: errors_paho.PahoConnectionRefusedBadUsernamePasswordError,
+    mqtt.CONNACK_REFUSED_NOT_AUTHORIZED: errors_paho.PahoConnectionRefusedNotAuthorizedError,
+}
+
+# mapping of Paho rc codes to Error object classes
+paho_rc_to_error = {
+    mqtt.MQTT_ERR_NOMEM: errors_paho.PahoNoMemoryError,
+    mqtt.MQTT_ERR_PROTOCOL: errors_paho.PahoProtocolError,
+    mqtt.MQTT_ERR_INVAL: errors_paho.PahoInvalidArgsError,
+    mqtt.MQTT_ERR_NO_CONN: errors_paho.PahoNoConnectionError,
+    mqtt.MQTT_ERR_CONN_REFUSED: errors_paho.PahoConnectionRefusedError,
+    mqtt.MQTT_ERR_NOT_FOUND: errors_paho.PahoNotFoundError,
+    mqtt.MQTT_ERR_CONN_LOST: errors_paho.PahoConnectionLostError,
+    mqtt.MQTT_ERR_TLS: errors_paho.PahoTLSError,
+    mqtt.MQTT_ERR_PAYLOAD_SIZE: errors_paho.PahoPayloadSizeError,
+    mqtt.MQTT_ERR_NOT_SUPPORTED: errors_paho.PahoNotSupportedError,
+    mqtt.MQTT_ERR_AUTH: errors_paho.PahoAuthError,
+    mqtt.MQTT_ERR_ACL_DENIED: errors_paho.PahoACLDeniedError,
+    mqtt.MQTT_ERR_UNKNOWN: errors_paho.PahoUnknownError,
+    mqtt.MQTT_ERR_ERRNO: errors_paho.PahoUnknownError,
+    mqtt.MQTT_ERR_QUEUE_SIZE: errors_paho.PahoQueueSizeError,
+}
+
+# mapping of Paho-specific errors to more generic errors
+paho_error_to_generic_error = {
+    errors_paho.PahoConnectionRefusedProtocolVersionError: errors.TransportError,
+    errors_paho.PahoConnectionRefusedIdentifierRejectedError: errors.TransportError,
+    errors_paho.PahoConnectionRefusedServerUnavailableError: errors.ConnectionFailedError,
+    errors_paho.PahoConnectionRefusedBadUsernamePasswordError: errors.UnauthorizedError,
+    errors_paho.PahoConnectionRefusedNotAuthorizedError: errors.UnauthorizedError,
+    errors_paho.PahoNoMemoryError: errors.TransportError,
+    errors_paho.PahoProtocolError: errors.TransportError,
+    errors_paho.PahoInvalidArgsError: errors.TransportError,
+    errors_paho.PahoNoConnectionError: errors.ConnectionDroppedError,
+    errors_paho.PahoConnectionRefusedError: errors.UnauthorizedError,
+    errors_paho.PahoNotFoundError: errors.TransportError,
+    errors_paho.PahoConnectionLostError: errors.ConnectionDroppedError,
+    errors_paho.PahoTLSError: errors.UnauthorizedError,
+    errors_paho.PahoPayloadSizeError: errors.TransportError,
+    errors_paho.PahoNotSupportedError: errors.TransportError,
+    errors_paho.PahoAuthError: errors.UnauthorizedError,
+    errors_paho.PahoACLDeniedError: errors.UnauthorizedError,
+    errors_paho.PahoUnknownError: errors.TransportError,
+    errors_paho.PahoQueueSizeError: errors.TransportError,
+}
+
+
+def _wrap_paho_error_in_generic_error(paho_error):
```

Comment:
>_wrap_paho_error_in_generic_error [](start = 4, length = 33)

open issue: I'd like feedback on this idea.  I think it makes sense.  For Paho errors, I convert the paho error code into a paho-specific error object (from errors_paho.py), and then I wrap that paho error inside a more generic error (from errors.py).  This way, the user gets a generic error, but they have the paho-specific details if they want them.

Reply:
i like the idea in general. But is it possible to see a stack trace with all the wrappings ? From the code here it seems it is mapping and returning the generic error.


---
In reply to: [302268054](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r302268054) [](ancestors = 302268054)